### PR TITLE
Integrate Fuzz Introspector with coverage build

### PR DESCRIPTION
**Pull request by Navidem**
Added build steps to the coverage build to integrate fuzz intrsopsector.

### CODE REVIEW

File: infra/build/functions/build_and_run_coverage.py [hunk 1]
```diff
@@ -223,6 +243,110 @@ def get_build_steps(  # pylint: disable=too-many-locals, too-many-arguments
...
+    #removes index.html from the end of url
+    coverage_url = bucket.html_report_url[:-11]
+    build_steps.extend(
+        get_fuzz_introspector_steps(project, project_name, base_images_project,
+                                    config, coverage_url))
+  return build_steps
+
+
+def get_fuzz_introspector_steps(project, project_name, base_images_project,
+                                config, coverage_url):
```
Comment (oliverchang) [inline infra/build/functions/build_and_run_coverage.py:258]: Please add a docstring.

File: infra/build/functions/build_and_run_coverage.py [hunk 2]
```diff
@@ -223,6 +243,110 @@ def get_build_steps(  # pylint: disable=too-many-locals, too-many-arguments
...
+  build_steps = []
+  FI_dir = '/workspace/fuzz-introspector/'
```
Comment (oliverchang) [inline infra/build/functions/build_and_run_coverage.py:260]: Use the full name. i.e. FUZZ_INTROSPECTOR_DIR, and put it above globally together with the other constants.

File: infra/build/functions/build_and_run_coverage.py [hunk 3]
```diff
@@ -223,6 +243,110 @@ def get_build_steps(  # pylint: disable=too-many-locals, too-many-arguments
...
+  oss_integration_dir = 'oss_fuzz_integration/'
```
Comment (oliverchang) [inline infra/build/functions/build_and_run_coverage.py:261]: Also, let's not hardcode '/' in the directory constants.

Use `os.path.join` where needed. In this case though, we can skip the separator completely and just do

```
oss_fuzz_integration_dir = 'oss_fuzz_integration'
```

File: infra/build/functions/build_and_run_coverage.py [hunk 4]
```diff
@@ -223,6 +243,110 @@ def get_build_steps(  # pylint: disable=too-many-locals, too-many-arguments
...
+  if project.fuzzing_language in ['c', 'c++']:
+    #removes index.html from the end of url
+    coverage_url = bucket.html_report_url[:-11]
+    build_steps.extend(
+        get_fuzz_introspector_steps(project, project_name, base_images_project,
+                                    config, coverage_url))
+  return build_steps
+
+
+def get_fuzz_introspector_steps(project, project_name, base_images_project,
```
Comment (oliverchang) [inline infra/build/functions/build_and_run_coverage.py:257]: Please write a test for this as well. See the existing build_and_run_coverage_test.py file for examples.

File: infra/build/functions/build_and_run_coverage.py [hunk 5]
```diff
@@ -223,6 +243,110 @@ def get_build_steps(  # pylint: disable=too-many-locals, too-many-arguments
...
+  build_steps.append(clone_step)
+
+  build_steps.append({
+      'name':
+          build_project.get_runner_image_name(base_images_project,
+                                              config.test_image_suffix),
+      'args': [
+          'bash', '-c',
+          (f'cd {FI_dir} && cd {oss_integration_dir}'
+           ' && sed -i \'s/\.\/infra\/base\-images\/all.sh/#\.\/infra\/base\-images\/all.sh/\''
```
Comment (oliverchang) [inline infra/build/functions/build_and_run_coverage.py:286]: This is very hacky and fragile. 

Let's upstream changes and do things in the main oss-fuzz repo where we can rather than using sed. 

Also is this actually still needed when we are building the customized build images in the base images build function?

File: infra/build/functions/build_and_run_coverage.py [hunk 6]
```diff
@@ -223,6 +243,110 @@ def get_build_steps(  # pylint: disable=too-many-locals, too-many-arguments
...
+      ],
+  })
+
+  build_steps.append(
+      build_project.get_compile_step(project, build, env, config.parallel))
+
+  #adjust coverage url
+  cov_url_escaped = coverage_url.replace("/", "\/").replace(":", "\:")
+  set_cov_url = (
+      f'sed -i \'s/http\:\/\/localhost\:8008\/covreport\/linux/{cov_url_escaped}/\''
```
Comment (oliverchang) [inline infra/build/functions/build_and_run_coverage.py:324]: Why is this needed? Good to explain with a comment in cases like this.

Also, is there a better way to do this that doesn't involve using sed?

File: infra/build/functions/build_project.py [hunk 7]
```diff
@@ -195,12 +195,18 @@ def get_env(fuzzing_language, build):
       'HOME': '/root',
       'OUT': build.out,
   }
+
```
Comment (oliverchang) [inline infra/build/functions/build_project.py:198]: nit: remove unnecessary change.

File: infra/build/functions/build_project.py [hunk 8]
```diff
@@ -195,12 +195,18 @@ def get_env(fuzzing_language, build):
...
   return list(sorted([f'{key}={value}' for key, value in env_dict.items()]))
 
 
 def get_compile_step(project, build, env, parallel):
   """Returns the GCB step for compiling |projects| fuzzers using |env|. The type
   of build is specified by |build|."""
+  set_git_repo_env = ''  #do nothing
```
Comment (oliverchang) [inline infra/build/functions/build_project.py:205]: no need for the comment here.

File: infra/build/functions/build_project.py [hunk 9]
```diff
@@ -195,12 +195,18 @@ def get_env(fuzzing_language, build):
...
+  if build.sanitizer == 'instrumentor':
+    set_git_repo_env = (
+        ' && export GITHUB_REPO=$(grep -P -o "\S+github.com\S+" /workspace/oss-fuzz/projects/'
```
Comment (oliverchang) [inline infra/build/functions/build_project.py:208]: Can we do this in a less hacky way?

It's used by the helper.py script right?

We can instead modify the helper.py script to set this based on the `main_repo` property in project.yaml. e.g. https://github.com/google/oss-fuzz/blob/126cf1128ca9a632fb882097730333d39418741a/projects/alembic/project.yaml#L6. Let's do all this in another PR which upstreams the patches from fuzz introspector.

This is slightly different from how the existing patch does it: https://github.com/ossf/fuzz-introspector/blob/a49f0ca54103e6dc0177700d22a166a727683334/oss_fuzz_integration/oss-fuzz-patches.diff#L194 but it's less hacky.

File: infra/build/functions/build_and_run_coverage.py [hunk 10]
```diff
@@ -223,6 +243,110 @@ def get_build_steps(  # pylint: disable=too-many-locals, too-many-arguments
       build_lib.http_upload_step(latest_report_info_body,
                                  latest_report_info_url,
                                  LATEST_REPORT_INFO_CONTENT_TYPE))
+
+  #currently fuzz introspector only supports c and c++
```
Comment (oliverchang) [inline infra/build/functions/build_and_run_coverage.py:247]: add a space after all '#'

File: infra/build/functions/build_and_run_coverage.py [hunk 11]
```diff
@@ -223,6 +243,110 @@ def get_build_steps(  # pylint: disable=too-many-locals, too-many-arguments
...
+  if project.fuzzing_language in ['c', 'c++']:
+    #removes index.html from the end of url
```
Comment (oliverchang) [inline infra/build/functions/build_and_run_coverage.py:249]: here too.

File: infra/build/functions/build_and_run_coverage.py [hunk 12]
```diff
@@ -223,6 +243,110 @@ def get_build_steps(  # pylint: disable=too-many-locals, too-many-arguments
...
+    coverage_url = bucket.html_report_url[:-11]
```
Comment (oliverchang) [inline infra/build/functions/build_and_run_coverage.py:250]: what is -11? either calculate this from a len(CONSTANT) or do a CONSTANT = 11

**Comment by oliverchang** [comment]
Also please fix the lint failures in https://github.com/google/oss-fuzz/runs/4622563959?check_suite_focus=true

**Comment by jonathanmetzman** [comment]
CI is failing because of test and presubmit failures. Please fix them.

File: infra/build/functions/build_and_run_coverage.py (hunk 5 above)
Comment (jonathanmetzman) [reply infra/build/functions/build_and_run_coverage.py:286]: +++1.
Also I don't like that this is relying on all.sh. As far as I know all.sh is meant for local development.

File: infra/build/functions/build_project.py [hunk 13]
```diff
@@ -195,12 +195,18 @@ def get_env(fuzzing_language, build):
...
   }
+
   return list(sorted([f'{key}={value}' for key, value in env_dict.items()]))
 
 
 def get_compile_step(project, build, env, parallel):
   """Returns the GCB step for compiling |projects| fuzzers using |env|. The type
   of build is specified by |build|."""
+  set_git_repo_env = ''  #do nothing
+  if build.sanitizer == 'instrumentor':
```
Comment (jonathanmetzman) [inline infra/build/functions/build_project.py:206]: I don't like that the sanitizer here is called instrumentor. That's a pretty ambiguous name given that sanitizers are instrumentation.
Why don't we call it introspector?

File: infra/build/functions/build_project.py (hunk 8 above)
Comment (jonathanmetzman) [inline infra/build/functions/build_project.py:205]: iirc what's going on here, it's hacky and we should do something else.
What I think you are doing is that when sanitizer is not "instrumentor" an empty string is inserted into the command on line 227.
But when it is "instrumentor" the command `&& export GITHUB_REPO...` is inserted.
IMO just change the entire command an each scenario, don't insert an empty string or `&& ...` that's really hard to understand.

File: infra/build/functions/build_and_run_coverage.py [hunk 14]
```diff
@@ -223,6 +243,110 @@ def get_build_steps(  # pylint: disable=too-many-locals, too-many-arguments
...
+                                config, coverage_url):
+  build_steps = []
+  FI_dir = '/workspace/fuzz-introspector/'
+  oss_integration_dir = 'oss_fuzz_integration/'
+
+  report_date = build_project.get_datetime_now().strftime('%Y%m%d')
+  bucket = IntrospectorBucket(project.name, report_date, PLATFORM,
+                              config.testing)
+
+  build = build_project.Build('libfuzzer', 'instrumentor', 'x86_64')
```
Comment (jonathanmetzman) [inline infra/build/functions/build_and_run_coverage.py:267]: How about you use the constants `FUZZING_ENGINE` and `ARCHITECTURE` here (like you did with `PLATFORM` instead of duplicating the strings?

File: infra/build/functions/build_and_run_coverage.py [hunk 15]
```diff
@@ -223,6 +243,110 @@ def get_build_steps(  # pylint: disable=too-many-locals, too-many-arguments
...
+  env = build_project.get_env(project.fuzzing_language, build)
+
+  clone_step = {
+      'args': [
+          'clone', 'https://github.com/ossf/fuzz-introspector.git', '--depth',
```
Comment (jonathanmetzman) [inline infra/build/functions/build_and_run_coverage.py:272]: @oliverchang Do we need to pin this?
Totally happy to ignore this for now.

File: infra/build/functions/build_and_run_coverage.py [hunk 16]
```diff
@@ -223,6 +243,110 @@ def get_build_steps(  # pylint: disable=too-many-locals, too-many-arguments
...
+  }
+  build_steps.append(clone_step)
+
+  build_steps.append({
+      'name':
+          build_project.get_runner_image_name(base_images_project,
+                                              config.test_image_suffix),
+      'args': [
+          'bash', '-c',
+          (f'cd {FI_dir} && cd {oss_integration_dir}'
```
Comment (jonathanmetzman) [inline infra/build/functions/build_and_run_coverage.py:285]: This is very difficult to understand.
Instead of defining `oss_integration_dir` to implicitly be a subdir of `FI_DIR` make it an absolute path like so:
```
oss_integration_dir = os.path.join(FI_DIR, 'oss_fuzz_integration')
```


And then you will only have to cd once here.

File: infra/build/functions/build_and_run_coverage.py (hunk 3 above)
Comment (jonathanmetzman) [inline infra/build/functions/build_and_run_coverage.py:261]: Could you explain what oss_integration_dir is in a comment?
I had to figure out what it is by going to the fuzz_introspector repo, can't expect readers to do this.

File: infra/build/functions/build_and_run_coverage.py [hunk 17]
```diff
@@ -223,6 +243,110 @@ def get_build_steps(  # pylint: disable=too-many-locals, too-many-arguments
...
+           ' && sed -i \'s/\.\/infra\/base\-images\/all.sh/#\.\/infra\/base\-images\/all.sh/\''
+           ' build_patched_oss_fuzz.sh'
+           ' && ./build_patched_oss_fuzz.sh')
```
Comment (jonathanmetzman) [inline infra/build/functions/build_and_run_coverage.py:288]: Where is this file?

File: infra/build/functions/build_and_run_coverage.py [hunk 18]
```diff
@@ -223,6 +243,110 @@ def get_build_steps(  # pylint: disable=too-many-locals, too-many-arguments
...
+  })
+
+  build_steps.append({
+      'name':
+          build_project.get_runner_image_name(base_images_project,
+                                              config.test_image_suffix),
+      'args': [
+          'bash', '-c',
+          ('sed -i s/base-builder/base-builder:introspector/g '
+           f'{FI_dir}{oss_integration_dir}oss-fuzz/projects/{project_name}/Dockerfile'
```
Comment (jonathanmetzman) [inline infra/build/functions/build_and_run_coverage.py:299]: OK, I think I figured out what this is doing and I think this isn't well designed.
If I understand correctly, the highlighted lines calls `build_patched_oss_fuzz.sh` from our cloned copy of fuzz-introspector which then [clones oss-fuzz again](https://github.com/ossf/fuzz-introspector/blob/main/oss_fuzz_integration/build_patched_oss_fuzz.sh#L23).
I think before we continue with this PR we should go back to the drawing board and describe at a high-level how this should work and then implement based on that spec.

File: infra/build/functions/build_and_run_coverage.py [hunk 19]
```diff
@@ -223,6 +243,110 @@ def get_build_steps(  # pylint: disable=too-many-locals, too-many-arguments
...
+  # to the lack of `-T` option (it creates a subdir in the destination dir).
+  build_steps.append(build_lib.gsutil_rm_rf_step(upload_report_url))
+  build_steps.append({
+      'name':
+          'gcr.io/cloud-builders/gsutil',
+      'args': [
+          '-m',
+          'cp',
+          '-r',
+          os.path.join(build.out, 'inspector-tmp'),
```
Comment (jonathanmetzman) [inline infra/build/functions/build_and_run_coverage.py:345]: What's `inspector-tmp`? Where does this come from?

**Review by jonathanmetzman — CHANGES_REQUESTED** [review]
I think this PR needs some redesigning at a high-level before we proceed on fixing individual issues.
I'm most concerned with the interface between oss-fuzz and fuzz-introspector (as is, this code (which lives in oss-fuzz) clones fuzz-introspector, which clones oss-fuzz again.
I'm also concerned with the frequent use of sed here. I think it probably doens't need to be used at all.

File: infra/build/functions/build_and_run_coverage.py (hunk 18 above)
Comment (DavidKorczynski) [reply infra/build/functions/build_and_run_coverage.py:299]: To follow on this, there are some other thoughts on fuzz-introspector that I think makes sense to consider when integrating fuzz-introspector to oss-fuzz: https://github.com/google/oss-fuzz/pull/7059#issuecomment-1006817896

File: infra/build/functions/build_and_run_coverage.py (hunk 15 above)
Comment (oliverchang) [reply infra/build/functions/build_and_run_coverage.py:272]: Yes, it would've been a good idea to pin this to the same hash used for the clang build. That said, we don't need this anymore after https://github.com/google/oss-fuzz/pull/7122 right?

Comment (Navidem) [reply infra/build/functions/build_and_run_coverage.py:272]: yes, won't need these steps.

**Comment by Navidem** [comment]
Thanks for all the comments, closing this as #7162 is the one to move forward with.
